src/tech/v3/dataset/column.clj

```diff
@@ -59,6 +60,53 @@
   (col-proto/set-missing col idx-seq))
 
 
+(defn index-structure
+  "Returns an index structure for the column. The index structure can help optimize
+  certain operations as they can provide faster lookup and subsetting of a column
+  or dataset.
+
+  Also see: tech.v3.dataset.column/select-from-index."
+  [col]
+  (col-proto/index-structure col))
+
+
+(defn with-index-structure
+  "Returns a copy of the column that will return an index-structure using the
+  provided `custom-make-index-strucutre-fn`. It also registers `klass` aliased
+  to `datatype-keyword` so it can be identified by the typing system."
+  [col datatype-keyword klass custom-make-index-structure-fn]
+  (col-proto/with-index-structure col datatype-keyword klass custom-make-index-structure-fn))
+
+
+(defn select-from-index
+  "Select a subset of the index as specified by the `mode` and
+  `selection-spec`. Each index-structure type supports different
+  modes. By default, tech.ml.dataset supports two index types:
+  `java.util.TreeMap` for non-categorical data, and `java.util.LinkedHashMap`
+  for categorical data.
+
+  Mode support for these types is as follows:
+
+  `java.util.TreeMap` - supports `::slice` and `::pick`.
+  `java.util.LinkedHashMap - support `::pick`
+
+  Usage by mode:
+
+  ::slice
+  (require '[tech.v3.dataset.column.column-index-structure :as col-index])
+  (select-from-index ::col-index/slice {:from 2 :to 5})
+  (select-from-index ::col-index/slice {:from 2 :from-inclusive? true :to 5 :to-inclusive? false})
+
+  ::pick
+  (require '[tech.v3.dataset.column.column-index-structure :as col-index])
+  (select-from-index ::col-index/pick [:a :c :e])
+
+  Note: Other index structures defined by other libraries may define other modes, so
+        consult documentation."
+  [index-structure mode selection-spec]
+  (col-proto/select-from-index index-structure mode selection-spec))
+
+
```

Comment:
Is this the right place to put these?

Reply:
Yes for sure with analogues at the dataset level defined in base.clj or in dataset.clj.  If the first argument is a column then my opinion is these should go or at least be accessible from column.clj.